### ISSUE TITLE
Paper #423 Understanding and Discovering Software Configuration Dependencies in Cloud and Datacenter Systems

### ISSUE THREAD

**Issue by kstolee** [researchart/fse20#89]
This issue is for discussion related to PR #38 

Author contact: @chenquingrong

**Comment by kstolee** [comment]
# Authors: 
Qingrong Chen (University of Illinois Urbana-Champaign)
Teng Wang (National University of Defense Technology)
Owolabi Legunsen (University of Illinois Urbana-Champaign)
Shanshan Li (National University of Defense Technology)
Tianyin Xu (University of Illinois Urbana-Champaign)

**Comment by art-reviewer** [comment]
[Paper](https://github.com/xlab-uiuc/fse20/blob/eba233ca7f23ecec745a085c92dd0303e5991a4e/submissions/reusable/paper423_cdep/esecfse2020-paper423.pdf)

**Comment by anon-eval** [comment]
@chenquingrong
Thank you for submitting the artifacts. Congratulations on the paper :)

**TLDR**: Based on the current version of your submission, I can only recommend the **available ** and **functional** badges but **not the reusable** badge. That said, I would be happy to update my recommendation if you can make the following changes
1. Please provide steps on how cDep can be used on a custom project. Does this require new configurations, additional dependencies e.t.c?
2. Please provide the optimal memory requirements for the docker image. I was running on a 16GB ubuntu 16.04 machine and I was only able to run it for `hdfs` and `mapreduce` datasets.
3. Please update `ReadMe.md` with the broken links.

I arrived at this conclusion using the following metrics
1. **Installation** ✅ : The installation runs seamlessly without any major errors. I used the docker image provided in the the github repository
2. **Public Availability**✅ : The tool is publicly available with a DOI
3. **Documentation** 🟡  : The tool is well documented for reproducing the results and executing cDep but not with respect to how the tool can be reused. Also the ReadMe files have some images which are not loaded and the authors use those images to explain their approach. Kindly fix these
4. **Consistent/Complete**✅ : The results are consistent on multiple executions and some of the results in the paper can be reproduced using the scripts provided although it is time taking. It took me 4 hours on a machine with 16Gb RAM
5. **Exercisable** ✅ : The tool is exercisable on the datasets provided in the repository
6. **Evidence of verification and validation**🟡 : I was able to only validate a portion of the results for `hdfs` and `mapreduce` datasets due to the large runtime and limited memory. I assume that this can be extended for other datasets as well.
7 **Reusability** ❌ : The authors should mention how cdep can be run for new projects not mentioned in this study. Does it need a certain configuration/dependencies/e.t.c?

Thanks :)

**Comment by AnonReviewer626** [comment]
@chenquingrong

## Available

Yes. Has a DOI and I could download the package from it.

## Functional

I am willing to grant this badge.

- *Documented*: The documentation is reasonably detailed.

- *Consistent*: The artifacts comprise the tool and results described in $6.

- *Complete*: It appears to be.

- *Exercisable*: It appears to be. See next remarks:

I cloned the repo. "mvn compile" ran without a problem.

I was able to initiate analysis using both "$ ./run.sh -a alluxio" and the Dockerized version (dockerrun.sh). The Docker container appears to start and do the same things as the un-Dockerized version.

The authors stated that it would take "tens of minutes" for the full run (I interpret this to mean some number less than an hour). I permitted the un-Dockerized analysis to run for 30 minutes on only alluxio. It did not finish. However, based on the report of @anon-eval I am willing to grant this badge.

## Reusable

I am not currently willing to grant this page. My concerns overlap somewhat with @anon-eval.

1. "Tens of minutes" is not very helpful. Indicate the machine specs you used and how long the analysis takes on (1) the fastest large-scale program, and  (2) on all of the programs together. That way I can estimate how things will go on my machine. Also, (3) add logging messages to indicate progress, e.g. "Starting task X…Performing sub-task A/Z, performing sub-task B/Z, …, Finished Task X, etc." I cannot tell if your program is stuck or making forward progress.

2. Indicate the resources required when running the un-Dockerized version -- #cores, RAM, any persistent storage needs, etc. During execution of "./run.sh -a alluxio" I saw core usage spike to all 8 of my cores and up to 4GB RAM. Are these limits typical for all analyzed programs? Do I need a bigger computer to run your analysis?

3. Indicate the fastest job to run so future users can "sanity check" their installation without waiting tens of minutes.

4. Several images in https://github.com/xlab-uiuc/cdep-fse are missing, e.g. the one that says "The following graph shows the end-to-end workflow". No image, so we cannot see the workflow!

5. Including a simple (but complete) program to analyze that can be easily understood would be helpful as an illustration. Section 5 of the README ("Verification and Validation") gives snippets but I want to see your tool work on an application containing these snippets. I think this is similar to @anon-eval's request for details about "How to run for new projects"?

6. I expect to see clear connections drawn between the results of the tool and the findings in the paper. Explain to me how your artifact produces the findings shown in the figures, sections, paragraphs, etc.

7. Your license file says "Copyright" but does not indicate who holds the copyright. Please clarify.

8. Please update the documentation to indicate the Java version required for compilation.

**Comment by AnonReviewer626** [comment]
@chenqingrong I think your handle was copied wrong earlier, I hope this sends you a notification.

**Comment by kstolee** [comment]
There seems to be agreement on the```available``` and ```functional``` badges. 

Thanks for the detailed reviews, @AnonReviewer626 and @anon-eval . 

@chenqingrong I suggest you update the artifact per the reviewer suggestions

**Comment by chenqingrong** [comment]
@anon-eval 
Thanks for your detailed reviews!

> 1. Please provide steps on how cDep can be used on a custom project. Does this require new configurations, additional dependencies e.t.c?

First, cDep could be well adapted for more applications and no more dependencies are needed for that. We implement one configuration interface as in https://github.com/xlab-uiuc/cdep-fse/blob/master/src/main/java/configinterface/ConfigInterface.java. The main reason is because different projects have different configuration interfaces to get configuration values. So, users just need to implement the interface for their own project such as https://github.com/xlab-uiuc/cdep-fse/blob/master/src/main/java/configinterface/HadoopInterface.java. Another thing they need to do is to download their own project source codes and put it under https://github.com/xlab-uiuc/cdep-fse/tree/master/app. The last thing they need to do is to add their application to the current command line interface in https://github.com/xlab-uiuc/cdep-fse/blob/62568ad2f2cb488b96a8243dbc77e4a4e17ef96a/src/main/java/cdep/cDep.java#L31.

Second, although not all projects need to do configuration dependency analysis, we believe the taint tracking of configuration parameters will benefit a lot of configuration related projects. The files https://github.com/xlab-uiuc/cdep-fse/blob/master/src/main/java/dataflow/DataTransform.java and https://github.com/xlab-uiuc/cdep-fse/blob/master/src/main/java/dataflow/IntraProcedureAnalysis.java implement all the logic for doing inter/intra data/control flow analysis for configuration parameters. The results of the taint tacking are stored in the object https://github.com/xlab-uiuc/cdep-fse/blob/62568ad2f2cb488b96a8243dbc77e4a4e17ef96a/src/main/java/dataflow/DataTransform.java#L132. So, if other projects want to analyze the results, they could just run the analysis, get the object and do the analysis they want. We use this object to do dependency analysis while it could be used for other analysis for sure.

These are my two main reasons for reusability and let me know if you have future questions:) Thanks!

> 2. Please provide the optimal memory requirements for the docker image. I was running on a 16GB ubuntu 16.04 machine and I was only able to run it for `hdfs` and `mapreduce` datasets.

Thanks for pointing this out. I was running it on a server with 62GB CentOS 7.8.2003 and it was ok for the whole datasets. I notice that if using docker image, 16GB might not be enough. If just using 16GB machine, one way I suggest is to use the third way **1.3 Building cDep in Your Own Environment** to run. It seems docker will restrict the memory a container could use before the system really kills that.

> 3. Please update `ReadMe.md` with the broken links.

Sorry for that. The old figures are from a private repo. I already fix that and submit a pull request.

Thanks again for your detailed reviews:)

**Comment by AnonReviewer626** [comment]
@chenqingrong 

> These are my two main reasons for reusability and let me know if you have future questions

Does your PR clarify these points? If not, please add them to the README.

**Comment by chenqingrong** [comment]
@AnonReviewer626 
Not yet. My PR currently only fixes broken links. I will update the README as well:)

**Comment by kstolee** [comment]
@chenqingrong please link your PRs in comments that reference them. (e.g., #144)

**Comment by chenqingrong** [comment]
@kstolee Thanks for the suggestion:)

@AnonReviewer626 The README changes are also merged. The related PR is #144 .

**Comment by chenqingrong** [comment]
@AnonReviewer626 
Thanks for your detailed reviews!
> 1. "Tens of minutes" is not very helpful. Indicate the machine specs you used and how long the analysis takes on (1) the fastest large-scale program, and  (2) on all of the programs together. That way I can estimate how things will go on my machine. Also, (3) add logging messages to indicate progress, e.g. "Starting task X…Performing sub-task A/Z, performing sub-task B/Z, …, Finished Task X, etc." I cannot tell if your program is stuck or making forward progress.

I am using the machine with CentOS 7.8.2003, 62GB memory and 40 Intel(R) Xeon(R) CPU E5-2660 v3 @ 2.60GHz. When I run the analysis, I use maven to directly build it and run it and it is faster than using the docker version. (1) The fastest program is **hdfs** and it takes around 15 minutes to finish. (2) On all programs, it takes around 3 hours. (3) I will add more logs to my codes and thanks for your suggestions.

> 2. Indicate the resources required when running the un-Dockerized version -- #cores, RAM, any persistent storage needs, etc. During execution of "./run.sh -a alluxio" I saw core usage spike to all 8 of my cores and up to 4GB RAM. Are these limits typical for all analyzed programs? Do I need a bigger computer to run your analysis?

I just try to run this on my local machine (not the server). My laptop is MacBook Pro with 16GB memory and 2.6 GHz 6-Core Intel Core i7. It finishes the analysis for **hdfs**  with 12:24 min. I would say if you do not see the message of OOM, then it is ok to finish the analysis on your current machine. However, bigger machine does finish early since it does not need to swap memory too frequently. The current logic for the programs is to read all jars from (app)[https://github.com/xlab-uiuc/cdep-fse/tree/master/app] and run the analysis. If you see the directory size (app)[https://github.com/xlab-uiuc/cdep-fse/tree/master/app], it is already 1.4GB.
 
> 3. Indicate the fastest job to run so future users can "sanity check" their installation without waiting tens of minutes.

One way is to try hdfs only and it should only take 10-15 minutes. Another way is that I already add one demo case section [5.6](https://github.com/xlab-uiuc/fse20/blob/master/submissions/available/paper423_cdep/README.md#56-demo-case) to the README and it should take less than 1 minute. Please pull the newest codes before you try this demo case. This demo case simply restrict the function cDep will analyze and it contains the case we describe in section [5.4](https://github.com/xlab-uiuc/fse20/blob/master/submissions/available/paper423_cdep/README.md#54-default-value-dependency).

> 4. Several images in https://github.com/xlab-uiuc/cdep-fse are missing, e.g. the one that says "The following graph shows the end-to-end workflow". No image, so we cannot see the workflow!

This is already fixed in the PR #144 .

> 5. Including a simple (but complete) program to analyze that can be easily understood would be helpful as an illustration. Section 5 of the README ("Verification and Validation") gives snippets but I want to see your tool work on an application containing these snippets. I think this is similar to @anon-eval's request for details about "How to run for new projects"?

I already add descriptions about how to run for new projects in the README file at Section 6. (PR #144)

> 6. I expect to see clear connections drawn between the results of the tool and the findings in the paper. Explain to me how your artifact produces the findings shown in the figures, sections, paragraphs, etc.

The findings in the figures include the results from our study part and our tool part. The findings from the study part are based on the [dataset](https://github.com/xlab-uiuc/cdep-fse/tree/master/dataset) we provide. The findings from the tool part are based on the [tool findings](https://github.com/xlab-uiuc/cdep-fse/tree/master/cDep_result). We are using the consistent terms in paper and in the dataset. So, any number in the paper could be reproduced just by counting in those datasets. We provide one script in 9b as well.

> 7. Your license file says "Copyright" but does not indicate who holds the copyright. Please clarify.

Addressed in PR #150.

> 8. Please update the documentation to indicate the Java version required for compilation.

 We already mention this under [1.3](https://github.com/xlab-uiuc/cdep-fse#13-building-cdep-in-your-own-environment): We build cDep using Java(TM) SE Runtime Environment (build 12.0.2+10) and Apache Maven 3.6.1.

> 9a. README says "It contains the following four files:" You list 5 files.

Addressed in PR #150.


> 9b. Similar to my item 6, please indicate some mapping between the dataset CSVs and the results in Table 7 (e.g. via a demonstrative `grep`, but preferably include your full analysis script that produces Table 7).

We do not have a full script when we do the counting and we simply rely on some grouping options from Excel to do that. We just write a simple [python script](https://github.com/xlab-uiuc/cdep-fse/blob/master/validate.py) to help count all the true positive cases found by cDep. The script is recently added, so pull before you run it.

**Comment by AnonReviewer626** [comment]
@chenqingrong Please ping here when #150 is landed and I will take another look.

**Comment by chenqingrong** [comment]
@AnonReviewer626 
Thank you! It is just landed. Let me know if you have more questions or concerns:)

**Comment by AnonReviewer626** [comment]
@chenqingrong Thank you for the changes, I have reviewed them.

My major concerns have been addressed. I am now willing to grant **available, functional, reusable**.

The missing images help a lot in understanding your approach and the additional examples look good. I was able to run the HDFS example and it took 15 minutes on my Macbook Pro.

However, I have several recommendations that would improve the artifact's reusability if the authors are so inclined. These are (almost) all minor documentation changes.

1. I urge the authors to document their machine specs in the paper and the artifact. I see the specs were given in this issue, but I cannot find them elsewhere. Since your paper includes running time numbers ("160 minutes"), you should include the machine specs. Running times on your nice server may be a lot faster than on a standard desktop or laptop.

2. I asked that you change the documentation to "Indicate the resources required when running the un-Dockerized version".

Again, thank you for noting this, but please put it into the artifact documentation. I do not see this information in the README or other artifact documentation.

3. You said "(3) I will add more logs to my codes and thanks for your suggestions."

The program still appears to be completely silent. I ran your example for 5.4 and got this output:

```
(09:03:05) jamie@woody ~/Desktop/fse20art/fse20/cdep-fse $ time mvn exec:java -Dorg.slf4j.simpleLogger.log.org.apache.maven.cli.transfer.Slf4jMavenTransferListener=warn -Dexec.mainClass="cdep.cDep" -Dexec.args="-o tmp.txt -x ./config_files -a yarn -demo"
[INFO] Scanning for projects...
[INFO] 
[INFO] -----------------------------< cDep:cDep >------------------------------
[INFO] Building cDep 1.0
[INFO] --------------------------------[ jar ]---------------------------------
[INFO] 
[INFO] --- exec-maven-plugin:3.0.0:java (default-cli) @ cDep ---
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  22.645 s
[INFO] Finished at: 2020-06-25T09:03:33-04:00
[INFO] ------------------------------------------------------------------------

real	0m23.658s
user	2m27.387s
sys	0m1.499s
```

The `dockerrun` script is also silent.

- Where do I look to see the result? I looked in `/tmp/output/cDep_result.csv` but it was not updated.
- Where can I find the demo file?

4. I wrote "Indicate the fastest job to run so future users can "sanity check" their installation without waiting tens of minutes."

Again, thank you for noting this in the issue, but I would like to see this in the artifact documentation.

5. README part 4 now says the modules and where to find them. Great!

I recommend that the authors consider documenting their code (at least on their next project :-D). There are 2KLoC of Java code here according to `cloc`. I see only two instances of Javadoc:

```
~/Desktop/fse20art/fse20/submissions/available/paper423_cdep/cdep-fse/src/main/java $ grep -R '/\*\*' .
./configinterface/HadoopInterface.java:  /**
./configinterface/HadoopInterface.java:  /**
```

I also see no evidence of unit tests or functional tests, which is surprising.

6. I like section $6 of the README. Step 1 is good, but could you add to it "See XYZ for an example config interface for HDFS" or similar? That way I can see what your use of this interface looks like. This is related to @anon-eval's final request from their initial review, I think.

-----

@anon-eval Do the changes satisfy you for **reusable** as well?

**Comment by chenqingrong** [comment]
@AnonReviewer626 
Thanks for your detailed reviews!

> 1. I urge the authors to document their machine specs in the paper and the artifact. I see the specs were given in this issue, but I cannot find them elsewhere. Since your paper includes running time numbers ("160 minutes"), you should include the machine specs. Running times on your nice server may be a lot faster than on a standard desktop or laptop.
> 2. I asked that you change the documentation to "Indicate the resources required when running the un-Dockerized version".

Already add them into the README (PR #156)

> 3. Where do I look to see the result? I looked in /tmp/output/cDep_result.csv but it was not updated.

I already provide a script `demo.sh` which will put the result at that place. Please pull the latest codes to try that.

> 3. Where can I find the demo file?

I do not create a new file for this demo. I just restrict the function name which will be analyzed under the demo mode as shown in [code](https://github.com/xlab-uiuc/cdep-fse/blob/042c8f109229c188e3aaa453b95eaced3030bc96/src/main/java/dataflow/DataTransform.java#L153). Any function which does not contain that name `shouldRMFailFast` won't be analyzed.

> 4. Again, thank you for noting this in the issue, but I would like to see this in the artifact documentation.

Already updated in PR #156 .

> 5. I also see no evidence of unit tests or functional tests, which is surprising. The lack of tests has implications for reusability as well -- if there aren't tests, how can I be sure that I haven't broken something important when I extend your code in some way?

Thanks for pointing it out. We ran out of time when we first developed this tool and just tried to make it able to run. 
I think you give great suggestions. We are considering to add tests for this project. However, it is just gonna take some time, so please stay tuned on our GitHub repo:)

> 6. I like section $6 of the README. Step 1 is good, but could you add to it "See XYZ for an example config interface for HDFS" or similar? That way I can see what your use of this interface looks like. 

I already update my descriptions in PR #156.

**Comment by anon-eval** [comment]
@chenqingrong Thank you for making these updates. I would also recommend the `reusable` badge.

**Comment by kstolee** [comment]
Thanks all for your involvement with the review process! 

@chenqingrong be sure to cite your artifact using the DOI in the references section of your paper, and congratulations!

**Comment by chenqingrong** [comment]
@AnonReviewer626 @anon-eval @kstolee 
Thank you all for the review process and thanks for all the precious suggestions:)
